AngularJS/src/org/angular2/inspections/AngularDuplicateAttributeInspection.java

```diff
@@ -0,0 +1,164 @@
+// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package org.angular2.inspections;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.angular2.codeInsight.attributes.Angular2AttributeDescriptor;
+import org.angular2.entities.Angular2Directive;
+import org.angular2.lang.Angular2Bundle;
+import org.angular2.lang.html.parser.Angular2AttributeType;
+import org.angular2.lang.selector.Angular2DirectiveSimpleSelector;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.intellij.codeInspection.ProblemsHolder;
+import com.intellij.lang.javascript.psi.JSElement;
+import com.intellij.psi.xml.XmlAttribute;
+import com.intellij.psi.xml.XmlElement;
+import com.intellij.psi.xml.XmlTag;
+
+import one.util.streamex.StreamEx;
+
+public class AngularDuplicateAttributeInspection extends AngularHtmlLikeTemplateLocalInspectionTool {
+   @Override
+   protected void visitXmlTag(
+         @NotNull final ProblemsHolder problemsHolder,
+         @NotNull final XmlTag xmlTag) {
+      StreamEx.of(xmlTag.getAttributes())
+              .mapToEntry(XmlAttribute::getDescriptor, Function.identity())
+              .selectKeys(Angular2AttributeDescriptor.class)
+              .filterKeys(AngularDuplicateAttributeInspection::isInspectable)
+              .flatMapKeyValue(AngularDuplicateAttributeInspection::toDuplicateAttribute)
+              .collect(Collectors.groupingBy(duplicateAttribute -> duplicateAttribute.targetName))
+              .values()
+              .forEach(duplicateAttributes -> {
+                 if (duplicateAttributes.size() > 1) {
+                    registerProblems(problemsHolder, duplicateAttributes);
+                 }
+              });
+   }
+
+   private static boolean isInspectable(@NotNull final Angular2AttributeDescriptor descriptor) {
+      final Angular2AttributeType type = descriptor.getInfo().type;
+      return type == Angular2AttributeType.REGULAR
+            || type == Angular2AttributeType.PROPERTY_BINDING
+            || type == Angular2AttributeType.BANANA_BOX_BINDING;
+   }
+
+   private static Stream<DuplicateAttribute> toDuplicateAttribute(
```

Comment:
Logic is slightly convoluted here and it's hard for me to follow it.

I think it would be simpler if you handle it as two separate cases. One would be redundant selector, which might actually be worth moving to a separate inspection and the other is duplicate binding.

In case of selectors you could group attrs by directives, from which originates selector. For each directive with multiple attributes representing redundant message would appear on those attributes who have single source of origin (selector) and are not bindings.

Reply:
Sounds reasonable to split the inspection!